OSDOCS-2230: Added OpenShift sandboxed containers 4.8 RN (#33105)
openshift/openshift-docs: release_notes/ocp-4-8-release-notes.adoc

=== Operator development

[id="ocp-4.8-sandboxed-containers"]
=== {sandboxed-containers-first}

> ==== {sandboxed-containers-first} support on {product-title} (Technology Preview)

Reviewer: Oops - missed that this also needs an id tag

> {sandboxed-containers-first} 1.0.0 Technology Preview release introduces users with built-in support for running Kata Containers as an additional runtime. {sandboxed-containers-first} provide users with the ability to choose from the type of workload that they want to run to cover a wide variety of use cases. You can use the {sandboxed-containers-operator} to perform tasks such as installation and removal, updates, and status monitoring. {sandboxed-containers-first} are only supported on bare metal. {op-system-first} is the only supported operating system for {sandboxed-containers-first} 1.0.0. Disconnected environments is not supported in {product-title} {product-version}.

Reviewer: ```suggestion
{sandboxed-containers-first} 1.0.0 Technology Preview release introduces built-in support for running Kata Containers as an additional runtime. {sandboxed-containers-first} provide users with the ability to choose from the type of workload that they want to run to cover a wide variety of use cases. You can use the {sandboxed-containers-operator} to perform tasks such as installation and removal, updates, and status monitoring. {sandboxed-containers-first} are only supported on bare metal. {op-system-first} is the only supported operating system for {sandboxed-containers-first} 1.0.0. Disconnected environments are not supported in {product-title} {product-version}.
```

Reviewer: Might also be worth breaking this blurb into two or more short paragraphs.

=== Operator development

[id="ocp-4.8-sandboxed-containers"]
=== {sandboxed-containers-first}

[id="ocp-4.8-sandboxed-containers-tp"]
==== {sandboxed-containers-first} support on {product-title} (Technology Preview)

> {sandboxed-containers-first} 1.0.0 Technology Preview release introduces built-in support for running Kata Containers as an additional runtime. {sandboxed-containers-first} provide users with the ability to choose from the type of workload that they want to run to cover a wide variety of use cases. You can use the {sandboxed-containers-operator} to perform tasks such as installation and removal, updates, and status monitoring.

Reviewer 2: "choose from the type of workload", I have a feeling there is a missing word here. Was the intent "choose the runtime based on the type of workload"? In other words, the question for me is "choose what"?

Author: @c3d "Choose the type of workload that they want..."? Maybe drop from?

Reviewer 2: I believe the same wording issue may exist in #32307

Reviewer 2: @zanetworker @jensfr Were you the ones that came up with the original wording.

It seems to me like we want to explain that we can chose the runtime class on a per-workload basis, but as written, I have no idea what this really says ;-)

Reviewer 3: @c3d nope. The sentence needs to be changed to reflect that we can choose the runtime based on the workload.

Reviewer 2: @zanetworker I'm confused, because you write "nope" and then repeat what I wrote, except you replace "runtime class on a per-workload basis" with "runtime based on the workload", so you seem to agree with me?

In any case, the way you change the runtime is by selecting a `runtimeClass`, so I believe that we should avoid introducing the notion of "runtime" here. WDYT?

Reviewer 3: >@zanetworker @jensfr Were you the ones that came up with the original wording.

@c3d the nope was meant to answer your first question.

Reviewer 3: >In any case, the way you change the runtime is by selecting a runtimeClass, so I believe that we should avoid introducing the notion of "runtime" here. WDYT?

At this point, I believe we should not get to the specific as kubernetes-specific term  vs. a technical reference.

Reviewer 2: @zanetworker I understand your point. And indeed, we only used "runtime" so far. Can you suggest a wording for @sagidlow to copy-paste?

Reviewer 3: OpenShift sandboxed containers provide users with the ability to choose Kata Containers as an additional runtime to provide additional isolation for their workloads. The OpenShift sandboxed containers operator is made available to users for performing tasks such as installation, removal, updates, and status monitoring of Kata Containers.

> {sandboxed-containers-first} are only supported on bare metal. {op-system-first} is the only supported operating system for {sandboxed-containers-first} 1.0.0. Disconnected environments are not supported in {product-title} {product-version}.

Reviewer 2: Why does `{product-title} {product-version}` show up as "OpenShift Container Platform Branch Build."? Are we making a support statement about OpenShift or about sandboxed containers?

Author: Branch Build will be 4.8. It's weird Preview link. When the doc goes live it will say 4.8.

=== Operator development

[id="ocp-4.8-sandboxed-containers"]
=== {sandboxed-containers-first}

[id="ocp-4.8-sandboxed-containers-tp"]
==== {sandboxed-containers-first} support on {product-title} (Technology Preview)

> {sandboxed-containers-first} 1.0.0 Technology Preview release introduces built-in support for running Kata Containers as an additional runtime. {sandboxed-containers-first} enables users to choose Kata Containers as an additional runtime to provide additional isolation for their workloads. The {sandboxed-containers-operator} enables users to perform tasks such as installation, removal, updates, and status monitoring of Kata Containers.

Reviewer 4: I am unclear how the operator allows status monitoring of Kata Containers.  I know of no specific feature for this.

Reviewer 3: I don't think it means `status` literally, but rather status (for example, if you describe KataConfig you could see the progress.

Author: > The OpenShift sandboxed containers Operator enables users to perform tasks such as installation, removal, updates, and monitoring status using the `KataConfig` custom resource.

Reviewer 3: Suggestion: The OpenShift sandboxed containers Operator enables users to perform tasks such as installation, removal, updates, and monitoring installation state visible when describing the KataConfig custom resource.

> {sandboxed-containers-first} 1.0.0 Technology Preview release introduces built-in support for running Kata Containers as an additional runtime. {sandboxed-containers-first} enables users to choose Kata Containers as an additional runtime to provide additional isolation for their workloads. The {sandboxed-containers-operator} enables users to perform tasks such as installation, removal, updates, and status monitoring of Kata Containers.

Reviewer 2: I watched this in preview mode, it displays as "OpenShift sandboxed containers 1.0.0 Technology Preview". What confuses me is that the only non-capitalized part here is the "most essential" part of the product name.

Could this be changed to show as "OpenShift Sandboxed Containers 1.0.0 Technology Preview"? Or alternatively, show the name in some different style?

Reviewer 3: @c3d the product name is OpenShift sandboxed containers

Reviewer 2: @zanetworker I understand that this is how we currently capitalize it. I still feel the need to point out that the following looks and reads wrong:

_OpenShift sandboxed containers 1.0.0 Technology Preview release introduces built-in support for running Kata Containers as an additional runtime. OpenShift sandboxed containers enables users to choose Kata Containers as an additional runtime to provide additional isolation for their workloads. The OpenShift sandboxed containers Operator enables users to perform tasks such as installation, removal, updates, and status monitoring of Kata Containers._

Maybe simply using a different font, e.g.:

_OpenShift sandboxed containers 1.0.0 Technology Preview release_ introduces built-in support for running _Kata Containers_ as an additional runtime. _OpenShift sandboxed containers_ enables users to choose _Kata Containers_ as an additional runtime to provide additional isolation for their workloads. The _OpenShift sandboxed containers Operator_ enables users to perform tasks such as installation, removal, updates, and status monitoring of _Kata Containers_.

Notice how the Parts in italicS Are totally Inconsistent in Their capitAlization.

Author: @c3d Discussions about this has happened previously. The approved product name is OpenShift sandboxed containers. Per our guidelines we have to capitalize our Operators. However, the approved product name is OpenShift sandboxed containers and OpenShift sandboxed containers Operator. Kata Containers is what I have found in their documentation in terms. It looks weird definitely, but this is what has been approved.

Reviewer 3: Which documentation are you referring to @sagidlow. I think Christoph's suggestion is trying to distinguish usage visually using italics as an example.

Reviewer 2: @zanetworker I suspect that @sagidlow is referring to how Kata Containers is spelled on their web site, e.g. https://katacontainers.io. It is rather consistently spelled with two capitals.

Frankly, even if that's something that needs to be revisited at the product management level, I think it's worth bringing up the issue that the current capitalisation makes documentation unreadable. In particular if we start capitalizing "operator"! It seriously takes me three or four back and forth scans in a sentence to simply be able to parse "OpenShift sandboxed containers 1.0.0 Technology Preview release introduces built-in support". To me, this is bad enough to turn into a UX experience at this point.

Different fonts could definitely help. Or the use of well-defined acronyms. If those are not options, then we need to pay attention to what the phrasing looks like in the end, something that may be a bit obscured to the writers because of the use of macros. For example 

_The release of the Technical Preview for OpenShift sandboxed containers 1.0.0 introduces built-in support for running Kata Containers as an additional runtime_

Author: @zanetworker @c3d - With our current guidelines we cannot use italics in this manner. We also cannot implement different fonts, and so on. 

We can talk about changing the name of the Operator or the product itself by using proper capitalization. But unfortunately, unless we can fix the product name. As @c3d this is def. a UX experience issue and if we can change OpenShift sandboxed containers to OpenShift Sandboxed Containers that would make more sense and more align with how our products are titled.

For more information, see xref:../authentication/managing_cloud_provider_credentials/cco-mode-sts.adoc#cco-mode-sts[Using manual mode with STS].

[id="ocp-4.8-sandboxed-containers"]
=== {sandboxed-containers-first}

[id="ocp-4.8-sandboxed-containers-tp"]
==== {sandboxed-containers-first} support on {product-title} (Technology Preview)

> {sandboxed-containers-first} 1.0.0 Technology Preview release introduces built-in support for running Kata Containers as an additional runtime. {sandboxed-containers-first} enables users to choose Kata Containers as an additional runtime to provide additional isolation for their workloads. The {sandboxed-containers-operator} enables users to perform tasks such as installation, removal, updates, and monitoring installation state when describing the `KataConfig` custom resource.

Reviewer 3: ```suggestion
{sandboxed-containers-first} 1.0.0 Technology Preview release introduces built-in support for running Kata Containers as an additional runtime. {sandboxed-containers-first} enables users to choose Kata Containers as an additional runtime to provide additional isolation for their workloads. The {sandboxed-containers-operator} automates the tasks of installing, removing, and updating Kata Containers. It allows for tracking the state of those tasks by describing the `KataConfig` custom resource.
```

Reviewer 3: One of the goals of the operator is to offload those tasks from the users.

registry.redhat.io/rhpam-7/rhpam-smartrouter-rhel8:7.9.0
----

[id="ocp-4.8-sandboxed-containers"]
=== {sandboxed-containers-first}

[id="ocp-4.8-sandboxed-containers-tp"]
==== {sandboxed-containers-first} support on {product-title} (Technology Preview)

{sandboxed-containers-first} 1.0.0 Technology Preview release introduces built-in support for running Kata Containers as an additional runtime. {sandboxed-containers-first} enables users to choose Kata Containers as an additional runtime to provide additional isolation for their workloads. The {sandboxed-containers-operator} enables users to perform tasks such as installation, removal, updates, and monitoring installation state when describing the `KataConfig` custom resource.

> {sandboxed-containers-first} are only supported on bare metal. {op-system-first} is the only supported operating system for {sandboxed-containers-first} 1.0.0. Disconnected environments are not supported in {product-title} {product-version}.

Reviewer 3: The version appears as `Branch Build` sometimes.

Author: @zanetworker `Branch Build` will result in the 4.8 product version showing up.

|TP
|

> |OpenShift sandboxed containers

Reviewer: There's not really a rhyme or reason to where things are added to this table, so if you want you could just stick it at the end since it's new. Here is fine too, though :+1:

Author: I'll leave it here for now, unless someone requests it to move it :-)

+
You can attempt to resolve the issue with the steps in this link:https://kb.vmware.com/s/article/2002779[VMware KBase article]. For more information, see the Red Hat Knowledgebase solution link:https://access.redhat.com/solutions/5785341[[UPI vSphere\] Node scale-up doesn't work as expected]. (link:https://bugzilla.redhat.com/show_bug.cgi?id=1918383[*BZ#1918383*])

* If you are using {sandboxed-containers-first}, you cannot use the hostPath volume in a {product-title} cluster to mount a file or directory from the host node’s file system into your pod.
+
(link:https://bugzilla.redhat.com/show_bug.cgi?id=1904609[*BZ#1904609*])

> * If you are running Fedora on {sandboxed-containers-first}, you need a workaround to install some packages. Some packages, like `iputils`, require file access permission changes that OSC does not grant to containers by default. To run containers that require such special permissions, it is necessary to add an annotation to the YAML file describing the workload, which tells `virtiofsd` to accept such file permissions for that workload. The required annotations are:

Reviewer: s/OSC/(I don't recall what the non-"first" attribute is, but it should go here I think :smiley:)

Author: @zanetworker - OSC here is intended to be OpenShift sandboxed containers?

Reviewer 3: @sagidlow I don't remember writing this, as I never use OSC :)

Maybe it's part of the initial sections in the overview that got moved to the release notes, we kind of were treating the RN section as a draft.

Reviewer 2: I wrote this part, but I don't recall writing OSC either (maybe I wrote "Kata"?)

Author: Was this maybe @jensfr ?

+
You can attempt to resolve the issue with the steps in this link:https://kb.vmware.com/s/article/2002779[VMware KBase article]. For more information, see the Red Hat Knowledgebase solution link:https://access.redhat.com/solutions/5785341[[UPI vSphere\] Node scale-up doesn't work as expected]. (link:https://bugzilla.redhat.com/show_bug.cgi?id=1918383[*BZ#1918383*])

> * If you are using {sandboxed-containers-first}, you cannot use the hostPath volume in a {product-title} cluster to mount a file or directory from the host node’s file system into your pod.

Reviewer 2: backtick `hostPath`

* If you are using {sandboxed-containers-first}, you cannot use the `hostPath` volume in a {product-title} cluster to mount a file or directory from the host node’s file system into your pod.
+
(link:https://bugzilla.redhat.com/show_bug.cgi?id=1904609[*BZ#1904609*])

* If you are running Fedora on {sandboxed-containers-first}, you need a workaround to install some packages. Some packages, like `iputils`, require file access permission changes that {product-title} does not grant to containers by default. To run containers that require such special permissions, it is necessary to add an annotation to the YAML file describing the workload, which tells `virtiofsd` to accept such file permissions for that workload. The required annotations are:
+
[source,yaml]
----
io.katacontainers.config.hypervisor.virtio_fs_extra_args: [ "-o", "modcaps=+sys_admin", "-o", "xattr" ]
----
+
(link:https://bugzilla.redhat.com/show_bug.cgi?id=1915377[*BZ#1915377*])

> * The Operator does not handle the custom `MachineConfigPool` which results in inconsistent behavior for installing and uninstalling the Operator.

Reviewer 4: This is just a feature that was not tested to our satisfaction.  It may work in some cases, but not all.  Just trying to set expectations.

Author: @cpmeadors How about something like?

> In the 4.8 release, the {sandboxed-containers-operator} does not handle the custom `MachineConfigPool`. Attempting to use customized `MachineConfigPool` may result in consistent behavior while installing and uninstalling the Operator. 

I try to steer clear of saying, well it might work in this instance but not in this...not sure. Any thoughts?

Reviewer 4: My thoughts exactly.  Looks good.

+
(link:https://issues.redhat.com/browse/KATA-764[*KATA-764*])

> * The Operator Hub details page about the {sandboxed-containers-operator} contains a few missing fields.

Reviewer 4: The reason for this to be included in the known issues was to reduce questions/support requests about completeness/readiness of the operator.  It is Tech Preview, but the incomplete information could indicate that it is unfinished more that to Tech Preview standards.  I don't know how much we have to explain, but that is why it is here.

Author: @cpmeadors In that case, I would clarify with something like this?

> The {sandboxed-containers-operator} details page on Operator Hub contains a few missing fields. The missing fields do not prevent you from installing the {sandboxed-containers-operator} in 4.8.

Reviewer 4: Works for me.

+
(link:https://issues.redhat.com/browse/KATA-826[*KATA-826*])

> * Creating multiple `KataConfig` custom resources results in a silent failure. The action itself seems to succeed, but does not provide the user with a failure prompt.

Reviewer 4: Technically, there is no failure here as sandboxed-containers-operator is unique (I think) in limiting the number of CRs created.  The issue here is how we handle a user creating additional CRs and setting correct expectations, Currently, people may assume they can create multiple CRs and then get confused when nothing changes.  Again, just background on the issue.

Author: @cpmeadors Should I clarify with something like this?

> Creating multiple `KataConfig` custom resources results in a silent failure.  The console does not provide a prompt to notify the user that creating more than one custom resource has failed. 

Let me know if this is not what you're looking for.

Reviewer 4: I like that wording.

* If you are using {sandboxed-containers-first}, you cannot use the `hostPath` volume in a {product-title} cluster to mount a file or directory from the host node’s file system into your pod.
+
(link:https://bugzilla.redhat.com/show_bug.cgi?id=1904609[*BZ#1904609*])

* If you are running Fedora on {sandboxed-containers-first}, you need a workaround to install some packages. Some packages, like `iputils`, require file access permission changes that {product-title} does not grant to containers by default. To run containers that require such special permissions, it is necessary to add an annotation to the YAML file describing the workload, which tells `virtiofsd` to accept such file permissions for that workload. The required annotations are:
+
[source,yaml]
----
io.katacontainers.config.hypervisor.virtio_fs_extra_args: [ "-o", "modcaps=+sys_admin", "-o", "xattr" ]
----
+
(link:https://bugzilla.redhat.com/show_bug.cgi?id=1915377[*BZ#1915377*])

> * In the 4.8 release, the {sandboxed-containers-operator} does not handle the custom `MachineConfigPool`. Attempting to use customized `MachineConfigPool` may result in consistent behavior while installing and uninstalling the Operator.

Reviewer 5: The problem here is that the actual feature (selecting nodes for the installation by adding a label to it) does work, but the UI that we present to the user is wrong. It can be worked around by adding the kataConfigPoolSelector manually. I suggest to add:

The impact of this is that the scheduling.nodeSelector will not be set, which means Pods using the
runtime class 'kata' could be scheduled to Nodes that don’t have the Kata Containers runtime installed.
This problem can be worked around by specifying the kataConfigPoolSelector manually in the kataconfig. In a terminal do 'oc edit kataconfig example-kataconfig' then add the selector.  A full example of a kataconfig CR

```
apiVersion: kataconfiguration.openshift.io/v1
kind: KataConfig
metadata:
  name: example-kataconfig
  spec:
    kataConfigPoolSelector:
      matchLabels:
        custom-kata1: test
```

* If you are using {sandboxed-containers-first}, you cannot use the `hostPath` volume in a {product-title} cluster to mount a file or directory from the host node’s file system into your pod.
+
(link:https://bugzilla.redhat.com/show_bug.cgi?id=1904609[*BZ#1904609*])

* If you are running Fedora on {sandboxed-containers-first}, you need a workaround to install some packages. Some packages, like `iputils`, require file access permission changes that {product-title} does not grant to containers by default. To run containers that require such special permissions, it is necessary to add an annotation to the YAML file describing the workload, which tells `virtiofsd` to accept such file permissions for that workload. The required annotations are:
+
[source,yaml]
----
io.katacontainers.config.hypervisor.virtio_fs_extra_args: [ "-o", "modcaps=+sys_admin", "-o", "xattr" ]
----
+
(link:https://bugzilla.redhat.com/show_bug.cgi?id=1915377[*BZ#1915377*])

> * In the 4.8 release, adding a value to `kataConfgPoolSelector` using the {product-title} web console results in `scheduling.nodeSelector` to be populated with an empty value. Pods using `RuntimeClass` with the value of `kata` may be scheduled to nodes that do not have the Kata Containers runtime installed.

Reviewer: ```suggestion
* In the 4.8 release, adding a value to `kataConfgPoolSelector` by using the {product-title} web console causes `scheduling.nodeSelector` to be populated with an empty value. Pods that use `RuntimeClass` with the value of `kata` might be scheduled to nodes that do not have the Kata Containers runtime installed.
```

+

> You can work around this issue by specifying the `nodeSelector` value manually in the `RuntimeClass` `kata` by running the following command:

Reviewer: ```suggestion
To work around this issue, specify the `nodeSelector` value manually in the `RuntimeClass` `kata` by running the following command:
```
just trying to avoid "by [gerund] by [gerund]" here :)
